image/layout/tar.go

```diff
@@ -0,0 +1,236 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package layout
+
+import (
+	"archive/tar"
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"time"
+
+	"github.com/opencontainers/image-spec/specs-go"
+	"golang.org/x/net/context"
+)
+
+// TarEntryByName walks a tarball pointed to by reader, finds an
+// entry matching the given name, and returns the header and reader
+// for that entry.  Returns os.ErrNotExist if the path is not found.
+func TarEntryByName(ctx context.Context, reader io.ReadSeeker, name string) (header *tar.Header, tarReader *tar.Reader, err error) {
+	_, err = reader.Seek(0, os.SEEK_SET)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	tarReader = tar.NewReader(reader)
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, nil, ctx.Err()
+		default:
+		}
+
+		header, err := tarReader.Next()
+		if err == io.EOF {
+			return nil, nil, os.ErrNotExist
+		}
+		if err != nil {
+			return nil, nil, err
+		}
+
+		if header.Name == name {
+			return header, tarReader, nil
+		}
+	}
+}
+
+// WriteTarEntryByName reads content from reader into an entry at name
+// in the tarball at file, replacing a previous entry with that name
+// (if any).  The current implementation avoids writing a temporary
+// file to disk, but risks leaving a corrupted tarball if the program
+// crashes mid-write.
+//
+// To add an entry to a tarball (with Go's interface) you need to know
+// the size ahead of time.  If you set the size argument,
+// WriteTarEntryByName will use that size in the entry header (and
+// Go's implementation will check to make sure it matches the length
+// of content read from reader).  If unset, WriteTarEntryByName will
+// copy reader into a local buffer, measure its size, and then write
+// the entry header and content.
+func WriteTarEntryByName(ctx context.Context, file io.ReadWriteSeeker, name string, reader io.Reader, size *int64) (err error) {
+	var buffer bytes.Buffer
+	tarWriter := tar.NewWriter(&buffer)
+
+	_, err = file.Seek(0, os.SEEK_SET)
+	if err != nil {
+		return err
+	}
+
+	tarReader := tar.NewReader(file)
+	found := false
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		var header *tar.Header
+		header, err = tarReader.Next()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return err
+		}
+
+		if header.Name == name {
+			found = true
+			err = writeTarEntry(ctx, tarWriter, name, reader, size)
```

Comment:
On Mon, Aug 29, 2016 at 07:08:33PM -0700, Stephen Day wrote:
> > +		if header.Name == name {
> > +			found = true
> > +			err = writeTarEntry(ctx, tarWriter, name, reader, size)
> 
> How does this avoid overwriting all subsequent entries in the tar file?

Because we're writing to a buffer [1] and only copy that buffer over
the file later [2].  More details on the implementation, the corners
it cuts, potential problems with it, and reasons why I went this way
anyway in the 3f19922d commit message.  And as I say there, folks
looking for a more reliable implementation should use a
filesystem-backed engine (which I haven't written up yet, but am happy
to add here if you like [3])

[1]: https://github.com/opencontainers/image-spec/pull/159/files#diff-745d1b74657ed8e87ac8c4e3f520f73aR77
[2]: https://github.com/opencontainers/image-spec/pull/159/files#diff-745d1b74657ed8e87ac8c4e3f520f73aR135
[3]: https://github.com/opencontainers/image-spec/pull/159#discussion_r76726544